sdk/storage/Azure.Storage.DataMovement/src/DownloadChunkHandler.cs

```diff
@@ -0,0 +1,281 @@
+                            {
+                                await _copyToDestinationFile(currentRange.Offset, (long) currentRange.Length, content).ConfigureAwait(false);
+                            }
+                            // Delete the temporary chunk file that's no longer needed
+                            File.Delete(chunkFilePath);
+                        }
+                        else
+                        {
+                            await InvokeFailedEvent(new FileNotFoundException(
+                                $"Could not append chunk to destination file at Offset: " +
+                                $"\"{currentRange.Offset}\" and Length: \"{currentRange.Length}\"," +
+                                $"due to the chunk file missing: \"{chunkFilePath}\"")).ConfigureAwait(false);
+                        }
+                    }
+                    catch
+                    {
+                        await InvokeFailedEvent(new Exception(
+                            $"Could not append chunk to destination file at Offset: " +
+                            $"\"{currentRange.Offset}\" and Length: \"{currentRange.Length}\""))
+                            .ConfigureAwait(false);
+                    }
+                }
+                else
+                {
+                    await InvokeFailedEvent(new ArgumentOutOfRangeException(
+                        "Offset",
+                        currentRange,
+                        $"Cannot find offset returned by Successful Download Range at Offset: " +
+                        $"\"{currentRange.Offset}\" and Length: \"{currentRange.Length}\""))
+                        .ConfigureAwait(false);
+                }
+
+                // Increment the current range we are expect, if it's null then
+                // that's the next one we have to wait on.
+                UpdateBytesAndRange((long)_ranges[_currentRangeIndex].Length);
+            }
+        }
+
+        private async Task InvokeFailedEvent(Exception ex)
+        {
+            foreach (HttpRange range in _ranges)
+            {
+                if (_rangesCompleted.TryRemove(range.Offset, out string tempChunkFile))
+                {
+                    if (File.Exists(tempChunkFile))
+                    {
+                        File.Delete(tempChunkFile);
+                    }
+                }
+            }
+            await _invokeFailedEventHandler(ex).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Updates Progress handler and the current Range we are waiting on
+        /// </summary>
+        /// <param name="bytesDownloaded"></param>
+        private void UpdateBytesAndRange(long bytesDownloaded)
+        {
+            Interlocked.Add(ref _bytesTransferred, bytesDownloaded);
+            _reportProgressInBytes(_bytesTransferred);
+            Interlocked.Increment(ref _currentRangeIndex);
```

Comment:
do `_bytesTransferred` and `_currentRangeIndex` have to be updated atomically?

Reply:
Yes, cause once an amount of bytes downloaded comes in, we have to update and it keep track of what range we're expected next.

Reply:
This doesn't update them atomically. You would need to add a lock around this method to accomplish that. As it's written now another thread can read these fields in between when `_bytesTransferred` and `_currentRangeIndex` are updated, essentially in a torn state.

Reply:
https://github.com/Azure/azure-sdk-for-net/issues/32853